src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,559 @@
+    // This results in lo = 01000000, but then the 6-th bit does not match,
+    // align up again gives us 10000000.
+    // We can align up directly to 10000000 by finding the first place after
+    // the highest mismatch such that both the corresponding bits are unset.
+    // Since all bits lower than the alignment are unset we don't need to
+    // align for the violations of ones anymore.
+    juint last_violation = W - 1 - count_leading_zeros(zero_violation);
+    T find_mask = std::numeric_limits<T>::max() << last_violation;
+    T either = lo | bits._zeros;
+    T tmp = ~either & find_mask;
+    T alignment = tmp & (-tmp);
+    lo = (lo & -alignment) + alignment;
+    return lo | bits._ones;
+  };
+
+  T new_lo = adjust_lo(bounds._lo, bits);
+  if (new_lo < bounds._lo) {
+    return {true, false, {}};
+  }
+  T new_hi = ~adjust_lo(~bounds._hi, {bits._ones, bits._zeros});
+  if (new_hi > bounds._hi) {
+    return {true, false, {}};
+  }
+
+  bool progress = (new_lo != bounds._lo) || (new_hi != bounds._hi);
+  bool present = new_lo <= new_hi;
+  return {progress, present, {new_lo, new_hi}};
+}
+
+// Try to tighten the known bit constraints from the bound information
+// E.g: if lo = 0 and hi = 10, then all but the lowest 4 bits must be 0
+template <class T>
+static AdjustResult<KnownBits<T>>
+adjust_bits_from_bounds(const KnownBits<T>& bits, const RangeInt<T>& bounds) {
+  static_assert(std::is_unsigned<T>::value, "");
+  T mismatch = bounds._lo ^ bounds._hi;
+  T match_mask = mismatch == 0 ? std::numeric_limits<T>::max()
+                               : ~(std::numeric_limits<T>::max() >> count_leading_zeros(mismatch));
+  T new_zeros = bits._zeros | (match_mask &~ bounds._lo);
+  T new_ones = bits._ones | (match_mask & bounds._lo);
+  bool progress = (new_zeros != bits._zeros) || (new_ones != bits._ones);
+  bool present = ((new_zeros & new_ones) == 0);
+  return {progress, present, {new_zeros, new_ones}};
+}
+
+// Try to tighten both the bounds and the bits at the same time
+// Iteratively tighten 1 using the other until no progress is made.
+// This function converges because bit constraints converge fast.
+template <class T>
+static NormalizeSimpleResult<T>
+normalize_constraints_simple(const RangeInt<T>& bounds, const KnownBits<T>& bits) {
+  AdjustResult<KnownBits<T>> nbits = adjust_bits_from_bounds(bits, bounds);
+  if (!nbits._present) {
+    return {false, {}, {}};
+  }
+  AdjustResult<RangeInt<T>> nbounds{true, true, bounds};
+  while (true) {
+    nbounds = adjust_bounds_from_bits(nbounds._data, nbits._data);
+    if (!nbounds._progress || !nbounds._present) {
+      return {nbounds._present, nbounds._data, nbits._data};
+    }
+    nbits = adjust_bits_from_bounds(nbits._data, nbounds._data);
+    if (!nbits._progress || !nbits._present) {
+      return {nbits._present, nbounds._data, nbits._data};
+    }
+  }
+}
+
+// Tighten all constraints of a TypeIntPrototype to its canonical form.
```

Comment:
Oh I like this definition of "canonical form". I think you should rename `normalize_constraints` -> `canonicalize`. Otherwise, you have two words, and you would need a definition for both "normalized" and "canonical".